Add the ability to draw hypergraphs as a bipartite graph

nwlandry opened the pull request:
This PR does the following:
* Removes `draw_dihypergraph` and moves its contents to `draw_bipartite`.
* The `draw_bipartite` now handles `Hypergraph` and `DiHypergraph` objects.
* Added the `bipartite_spring_layout` position function.

maximelucas commented:
Just a thought: conceptually, could this function just be the same as the `draw_dihypergraph` with the edge markers, except that the links are drawn without arrows and the layout would be different?

nwlandry commented:
Mmmmm. This is a great point. I think you're right. Would it make sense to group them together, call it `draw_bipartite` and have different handling for directed and undirected hypergraphs?

maximelucas commented:
I thought that could be a possibility. Worth thinking about. The current dihypergraph one only make sense as a bipartite if the edge markers are drawn though, I'd say.

Maybe directedness and "bipartite-ness" are two somewhat independent elements here.
Could the arrow drawing be included as a `directed` option in `draw_hyperedges`? Not sure that would include the edge markers. 

And then the `draw_bipartite` could just be about the layout really. Either the current layout we have in the  dihypergraph one, or a more bipartitey layout, the user could choose. And draw directed or undirected edges.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -5,7 +5,6 @@
 import matplotlib as mpl
 import matplotlib.pyplot as plt
 import numpy as np
-import seaborn as sb  # for cmap "crest"
```
I think we need to keep this even if it's "unused", because without it I was getting errors for using the cmap "crest" which is only provided by seaborn (and not matplotlib)

nwlandry commented:
I have now implemented all of your suggestions, @maximelucas. Should be ready to re-review.

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<p>The output was <code>(ax, node_coll, edge_coll)</code> and is now <code>(ax, node_coll, edge_coll, dyad_coll) </code>if I got that right. But the dyad_coll is a list of <code>FancyArrowPatch</code> and not an actual python <code>Collection</code>. </p><p> </p><p>The reason to output the collection was (i) for easy colorbars and (ii) to easily change parameters of the whole collection (like alpha value). What use did you have in mind for outputting the new list? Maybe we could make it an actual collection instead?</p><p> </p><p>I think my initial thought for not needing to output is that I thought either the arrows would be multicolored but consistent with the edges, in which case the edge_coll is sufficient, or be a single neutral color like black, in which case maybe returning them is less useful.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

maximelucas replied:
<p>Yea I was thinking maybe we don't return the list of patches. In any case I dont see a case where we would need a different colorbar for edge markers and bipartite edges. Maybe we enforce that if bipartite edges need to be colored with a cmap, it needs to be the same as that of edges markers?</p>

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<p>This case looks pretty bad now. How is it different from before? Should we not display it if it looks that bad?</p><p> </p><p>Also I thought you mentioned to change the pos into a single tuple argument, but maybe it's just the notebook that's not up to date, I haven't checked everything yet.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

nwlandry replied:
This is now removed and the `pos` issue is fixed.

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<p>same here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

nwlandry replied:
This example is now removed.

maximelucas commented on docs/source/api/drawing/xgi.drawing.draw.rst:
```diff
@@ -8,7 +8,6 @@ xgi.drawing.draw
    
    .. autofunction:: draw
-   .. autofunction:: draw_dihypergraph
    .. autofunction:: draw_multilayer
```
draw_bipartite should be added?

nwlandry replied:
Fixed. Good catch!

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
-        * "max_lines_lw" (default: 50)
-        * "min_source_margin" (default: 0)
-        * "min_target_margin" (default: 0)
+        * min_node_size
```
add default values like before

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
+        * max_edge_marker_size
+        * min_edge_marker_lw
+        * max_edge_marker_lw
+        * node_fc_cmap
```
not needed here, already a input argument

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
+        * min_edge_marker_lw
+        * max_edge_marker_lw
+        * node_fc_cmap
+        * node_ec_cmap
```
not used, is it?

nwlandry replied:
You're right! Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
+        * max_edge_marker_lw
+        * node_fc_cmap
+        * node_ec_cmap
+        * edge_marker_fc_cmap
```
not needed here, already a input argument

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
+        * node_fc_cmap
+        * node_ec_cmap
+        * edge_marker_fc_cmap
+        * edge_marker_ec_cmap
```
i think we don't use this for now

nwlandry replied:
Correct. Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
+        * edge_marker_ec_cmap
+        * min_dyad_lw
+        * max_dyad_lw
+        * dyad_color_cmap
```
not needed here, already a input argument

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1658,45 +1681,86 @@ def draw_dihypergraph(
+        "max_edge_marker_size": 5,
+        "min_dyad_lw": 1,
+        "max_dyad_lw": 5,
+        "node_fc_cmap": cm.Reds,
+        "node_ec_cmap": cm.RdBu,
+        "dyad_color_cmap": cm.Blues,
+        "edge_marker_fc_cmap": cm.Greys,
+        "edge_marker_ec_cmap": cm.Blues,
```
already in arguments above, or not used i think

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1658,45 +1681,86 @@ def draw_dihypergraph(
+        "max_node_lw": settings["max_node_lw"],
+        "min_node_size": settings["min_node_size"],
+        "max_node_size": settings["max_node_size"],
+        "node_fc_cmap": settings["node_fc_cmap"],
+        "node_ec_cmap": settings["node_ec_cmap"],
```
see above

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1658,45 +1681,86 @@ def draw_dihypergraph(
+        "max_node_lw": settings["max_edge_marker_lw"],
+        "min_node_size": settings["min_edge_marker_size"],
+        "max_node_size": settings["max_edge_marker_size"],
+        "node_fc_cmap": settings["edge_marker_fc_cmap"],
+        "node_ec_cmap": settings["edge_marker_ec_cmap"],
```
see above

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1707,111 +1771,154 @@ def draw_dihypergraph(
 
-    # draw "edge" phantom nodes
-    if edge_marker_toggle:
+    if node_labels is not None:
+        draw_node_labels(
+            H,
+            pos=node_pos,
+            node_labels=node_labels,
+            ax_nodes=ax,
+            zorder=3,
+        )
```
this can be passed simply to draw_nodes like before, right?

nwlandry replied:
Yep! Good catch!

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1707,111 +1771,154 @@ def draw_dihypergraph(
-            zorder=2,
-            params=settings,
-            rescale_sizes=rescale_sizes,
+    if hyperedge_labels is not None:
+        draw_node_labels(
+            H.dual(),
+            pos=edge_pos,
+            node_labels=hyperedge_labels,
+            ax_nodes=ax,
+            zorder=3,
```
same question

nwlandry replied:
Good catch! You're right, we can just use the `draw_nodes` function.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1707,111 +1771,154 @@ def draw_dihypergraph(
+
+
+def _draw_directed_edges(
+    edge_ids,
```
maybe second after ax, don't remember what we do in other functions

nwlandry replied:
Fixed. I deleted this arg and added `max_order` as an arg.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1707,111 +1771,154 @@ def draw_dihypergraph(
+    edge_ids,
+    ax,
+    H,
+    settings,
```
is usually last it ink

nwlandry replied:
Fixed.

nwlandry commented:
Thanks for the suggestions, @maximelucas! It looks like I need to take more time to address these concerns, and I'll let you know when I'm done!

nwlandry commented on xgi/drawing/draw.py:
```diff
@@ -1608,47 +1591,87 @@ def draw_dihypergraph(
         pairs.  By default, False.
+    max_order : int, optional
+        Maximum of hyperedges to plot. If None (default), plots all orders.
+    arrowsize : int (default=10)
```
Fixed.

nwlandry commented:
@maximelucas, I believe I addressed all of your comments! Thanks for such a thorough and helpful review! I did the following:
* I standardized all of the function arguments in response to your review.
* I made `draw_undirected_dyads` and `draw_directed_dyads` standalone functions and made them available to the user.
* Added default values to the settings and made them match other functions.

maximelucas commented on tests/drawing/test_draw.py:
```diff
@@ -663,7 +673,46 @@ def test_draw_bipartite_with_str_labels_and_isolated_nodes():
+    plt.close("all")
+
+
+# 1922, 1925, 1928-1929, 1936, 1942, 1949-1952, 1956, 1960, 2101, 2104, 2111, 2117, 2124-2126, 2130, 2134, 2163, 2166, 2173, 2189, 2199, 2210, 2236, 2247
```
is this needed?

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1978,13 +1986,7 @@ def draw_undirected_dyads(
-        [(node_pos[e[0]], edge_pos[e[1]]) for e in dyads if e[1] in edge_ids]
-    )
-
+    print(len(dyad_pos), len(dyad_color))
```
remove?

nwlandry replied:
Oops! Thanks for finding this!

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1947,24 +1936,43 @@ def draw_undirected_dyads(
+                ),
+            )
+        )
+
```
what does this do? Multiplies the number of elements of linewidth by the size of each edge?

nwlandry replied:
This allows you to specify a color for each edge ID and automatically apply it to all associated bipartite edges instead of specifying a color for all bipartite edges.

nwlandry replied:
I added a comment in the code explaining what these lines do.

maximelucas commented:
Thanks! Those are quick small comments for the last changes, I'll have a better look at the entire code soon.

maximelucas commented:
I'll do that next week!

nwlandry commented:
Great! Thanks so much!

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    xgi.draw_bipartite(DH)
</pre><p>The default coloring of hyperedges is a single color, instead of the crest cmap mapping edge size that we have in other functions. Any reason for this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

nwlandry replied:
Fixed!

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<p>The default cmap for hyperedges is now "Blues", which is different from all other functions where it is "crest". Any reason for that? I would maybe keep "crest" for consistency.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

maximelucas commented on docs/source/api/tutorials/In Depth 3 - Drawing DiHypergraphs.ipynb:
```diff
@@ -13,7 +13,7 @@
    "id": "ea8aad6a",
```
<p>is there a way to have the old "iterations=0" when wanted?</p><p>If not, maybe open an issue to re-implent it.</p><p>I like that in small cases, that one had the edge_markers at the barycenters. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xgi-org/xgi/pull/492/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

nwlandry replied:
I added a function called `edge_positions_from_barycenters` that takes in a dictionary of nodal positions and then returns a dictionary of edge positions computed from the barycenters.

nwlandry replied:
This is good because it is no longer tied to the specific position calculation (before, it was spring layout). Now you can use any of the node position methods (circular, spiral, Kamada-Kawai, etc.) to get edge positions from the barycenters.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1578,83 +1589,116 @@ def draw_dihypergraph(
-        After that, a spring layout is ran starting from the
-        initial layout, and each iteration makes all of the nodes overlap less.
+    aspect : {"auto", "equal"} or float, optional
+        Set the aspect ratio of the axes scaling, i.e. y/x-scale. `aspect` is passed
+        directly to matplotlib's `ax.set_aspect()`. Default is `equal`. See full
+        Set the aspect ratio of the axes scaling, i.e. y/x-scale. `aspect` is passed
+        directly to matplotlib's `ax.set_aspect()`. Default is `equal`. See full
+        description at
+        https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_aspect.html
```
```suggestion
        Set the aspect ratio of the axes scaling, i.e. y/x-scale. `aspect` is passed
        directly to matplotlib's `ax.set_aspect()`. Default is `equal`. See full
        description at
        https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_aspect.html
```

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1664,160 +1708,465 @@ def draw_dihypergraph(
+    Raises
+    ------
+    XGIError
+        If something different than a DiHypergraph is passed.
```
```suggestion
        If something different than a Hypergraph is passed.
```

maximelucas replied:
But I don't see that it raises an error, does it?

nwlandry replied:
Good catch! Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1578,83 +1589,116 @@ def draw_dihypergraph(
+            Collection containing the edges
 
     Raises
     ------
```
Below:

If something different than a Hypergraph is passed.

nwlandry replied:
Fixed.

maximelucas commented:
Some more minor comments.. we're almost there, it's a lot of code 😄 I'll check the rest of the code later today.

nwlandry commented:
@maximelucas thanks so much! Let me know when you're done and I'll start addressing your comments!

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1664,160 +1708,465 @@ def draw_dihypergraph(
+    if not pos:
+        pos = bipartite_spring_layout(H)
 
-    # set default colors
-    if lines_fc is None:
-        lines_fc = H_conv.edges.size
+    node_pos, edge_pos = pos
```
Since the `pos` format is different from other functions, should we raise an explicit error after checking if it's a two-dict tuple? Right now a single dict would give a generic error like "too much to unpack"

nwlandry replied:
I added a case checking this.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1664,160 +1708,465 @@ def draw_dihypergraph(
+    }
+    settings.update(kwargs)
+
+    if not pos:
+        pos = bipartite_spring_layout(H)
```
same

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1664,160 +1708,465 @@ def draw_dihypergraph(
+    }
+    settings.update(kwargs)
 
+    if not pos:
```
same

nwlandry replied:
Fixed.

maximelucas commented on xgi/drawing/draw.py:
```diff
@@ -1664,160 +1708,465 @@ def draw_dihypergraph(
+    settings.update(kwargs)
 
+    if not pos:
+        from ..convert import to_hypergraph
```
is this one purpose?

nwlandry replied:
To compute the layout? If so, yes.

maximelucas commented:
I think I checked all of it 😛 
I would just add an example somewhere of an undirected one, and raise an issue to create a new function with the "classic" graph bipartite layout, and maybe also with the barycenter layout. 
Thanks for all the work!

nwlandry commented:
@maximelucas, thanks for the review! A quick overview of my changes based on your comments:
* changed all of the dyads to be colored by the edge size (with "crest_r") by default
* Added error handling in a few places
* Added a new function `edge_positions_from_barycenters` as a more general replacement of the old functionality.
* Added an example of using `draw_bipartite` on an undirected hypergraph in Tutorial 5
* Added a few examples of the new `edge_positions_from_barycenters` in Tutorial 5 and In Depth 3.

At this point, I believe that I've addressed all your comments. Let me know if there are other things you would like me to change.
